### PR TITLE
Command user lock

### PR DESCRIPTION
**Pull request by taOsemc**
# PULL REQUEST

## **Overview**

CommandUserLock. Prevents a user from initiating several instances of the same command.

## **Status**

- [X] Typings have been updated or don't need to be.
- [X] This PR has been tested and is ready to be merged.

## **Semantic versioning classification**

- [x] MINOR: This PR adds new features, improves the code, and/or implies minimal API changes.

### CODE REVIEW

File: src/Core/Command/CommandUserLock.js [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+/**
+ * CommandUserLock. Prevents a user from initiating several instances of the same command.
+ *
+ * @author taO
+ *
+ * @class CommandUserLock
+ *
+ * @prop {Command} _command - The base command
+ * @prop {Set} _usersLocked - Set of user IDs
```
Comment (Khaaz) [inline src/Core/Command/CommandUserLock.js]:
```suggestion
 * @prop {Set<String>} _usersLocked - All user IDs currently locked
```

File: src/Core/Command/CommandUserLock.js [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+     * @type {Boolean} Options of userLock.
+     * @memberof CommandUserLock
+     */
+    get userLock() {
+        return this._command.options.userLock;
+    }
+
+    /**
+     * Checks if the user is locked.
+     *
```
Comment (Khaaz) [inline src/Core/Command/CommandUserLock.js]:
```suggestion
     * Lock the user if not locked.
```

File: src/Core/Command/CommandUserLock.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+     * @param {String} userID - The userID
+     * @returns {Boolean} Whether the user is unlocked or not.
+     *
+     * @memberof CommandUserLock
+     */
+    unLock(userID) {
+        const lock = this._usersLocked.has(userID);
+
+
```
Comment (Khaaz) [inline src/Core/Command/CommandUserLock.js]: remove the additional spacing here

**Review by Khaaz — CHANGES_REQUESTED** [review]
Looks good except some documentation changes.
Please update the typings as well. You can basically copy paste CommandCooldown for instance and start from here.

We might also want some messages the same way invalid cooldown, invalid permissions or invalid usage works.

File: src/Core/Command/CommandUserLock.js [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+     * Creates an instance of CommandUserLock.
+     *
+     * @param {Command} command
+     * @memberof CommandUserLock
+     */
+    constructor(command) {
+        this._command = command;
+
+        /**
+         * @type {Set<string>}>}
```
Comment (Khaaz) [inline src/Core/Command/CommandUserLock.js]:
```suggestion
         * @type {Set<string>}
```

**Review by Khaaz — APPROVED** [review]
lgtm, if @bsian03 can have a look 👀

File: types/Core/Command/CommandOptions.ts [hunk 5]
```diff
@@ -22,7 +22,7 @@ export declare class CommandOptions implements ACommandOptions {
 
     public deleteCommand?: boolean;
     public hidden?: boolean;
-
+    public userLock?: boolean;
```
Comment (bsian03) [inline types/Core/Command/CommandOptions.ts:26]: Keep the newline here, was added for organisation

**Review by bsian03 — CHANGES_REQUESTED** [review]
Just a minor thing, that's all

**Review by Khaaz — APPROVED** [review]

**Pull request merged**